### DISCUSSION TITLE
Pzem 017 shunt settings in tasmota

### DISCUSSION THREAD

**Discussion by grappini** [arendst/Tasmota#13733] [Ideas]
I use pzem 017 with a 200A shunt in tasmota. The current readings shows only half of the real value. According to the specifications different shunts need different setting. Is there any way to change this in tasmota? 

Register address | Description | Resolution
0x0003 | The current range | 
0x0000: 100A
0x0001: 50A
0x0002: 200A
0x0003: 300A

![pzemshuntsettings](https://user-images.githubusercontent.com/43513944/142690219-54a6784f-85c0-460d-a78c-92b1aa3ecca9.jpg)

**Comment by barbudor** [comment]
Hi
As of today this is not implemented
However it should be possible to implement this based on the EnergyConfig command.
I can have a look over the week-end but I will need your help for testing.

**Reply by grappini** [reply]
Sure. But please post the code  you implement. I want to understain the idea behind the code.

**Comment by grappini** [comment]
Any news regarding this ? I would try to fix it myself (as a newbie) but i don't know in which .ino file are the commands that read pzem 017 parameters and then change the current reading by  multiply   x2. It would be enough for me.

**Reply by Dlzensolar** [reply]
I encountered the same situation. The current value is doubled compared to the actual value. How to edit it?. I use tasmota 13.1.0. Thanks

**Reply by barbudor** [reply]
You need to co figure your PZEM in accordance to your shunt model : 50, 100 or 300A
Then it will send to Tasmota the right values.
Tasmota is making no calculation on the values sent by the PZEM but use them as-is

I introduced a Tasmota command to send the proper PZEM configuration command à couple of months ago. I thought it was documented...
I'm on a mobile for now. I'll search later

**Reply by barbudor** [reply]
Found it. Check the description for Energy Command here https://github.com/arendst/Tasmota/pull/19402#issue-1868620621

**Comment by grappini** [comment]
Changing in xnrg_06_pzem_dc.ino this line will do the trick ?

	Energy.current[PzemDc.channel] = (float)((buffer[5] << 8) + buffer[6]) / 100.0;  
	Energy.current[PzemDc.channel] = (float)((buffer[5] << 8) + buffer[6]) / 100.0 * 2; 

Any help from a dev would be appreciated. I'm not lazy but i don't know c++ and cannot fully understand tasmota structure because i read the code partialy blind.

 By doing so the energy counters will work correctly with the new modified amperage ?

**Reply by barbudor** [reply]
I didn't had enough time but the good solution is not to patch the calculation but to configure correctly the PZEM
I started to looks at how to send commands to the PZEM
You can do that in the meantime but that's not nice at all

**Reply by grappini** [reply]
I know it's an ugly patch (it would be nice to function though) but considering my condition it's hard to get anything better from me. I know basically very little of tasmota structure due to my lack of C++ programing.

**Comment by barbudor** [comment]
@grappini Please find below a test version.
It allows to configure the Current Range of the PZEM 017 using command `EnergyConfig`
The syntax is:
`EnergyConfig<x> range 50|100|200|300`
where <x> is the channel if you have more than 1 PZEM (if not used, will default to 1)
Exemple logs you should have when typing `EnergyConfig Range 50`
```
20:24:28.428 CMD: EnergyConfig Range 50
20:24:28.432 PDC: EnergyConfig1 Range 50
20:24:28.433 PDC: range_in=50
20:24:28.434 PDC: ModBus->Send(@:1, 0x06, 0x03, 1)
20:24:28.448 MQT: stat/nodemcu/ENERGYCONFIG = {"EnergyConfig":"Done"}
```

For now there is quite a few logs that you can't disable for debugging.
I don't think we need to save the settings in Flash. I would assume that the PZEM store that on it's own like the module address.
Please confirm if that works and if the settings are persistent.

tasmota.bin and tasmota.bin.gz: (obsolete binary removed) => Now feature is in Tasmota : see https://github.com/arendst/Tasmota/pull/19402

My development branch if you want to see the code: Now code is merged in Tasmota

**Reply by grappini** [reply]
Thanks , i will test it later on today.

**Reply by ortegafernando** [reply]
> @grappini Please find below a test version. It allows to configure the Current Range of the PZEM 017 using command `EnergyConfig` The syntax is: `EnergyConfig<x> range 50|100|200|300` where is the channel if you have more than 1 PZEM (if not used, will default to 1) Exemple logs you should have when typing `EnergyConfig Range 50`
> 
> ```
> 20:24:28.428 CMD: EnergyConfig Range 50
> 20:24:28.432 PDC: EnergyConfig1 Range 50
> 20:24:28.433 PDC: range_in=50
> 20:24:28.434 PDC: ModBus->Send(@:1, 0x06, 0x03, 1)
> 20:24:28.448 MQT: stat/nodemcu/ENERGYCONFIG = {"EnergyConfig":"Done"}
> ```
> 
> For now there is quite a few logs that you can't disable for debugging. I don't think we need to save the settings in Flash. I would assume that the PZEM store that on it's own like the module address. Please confirm if that works and if the settings are persistent.
> 
> tasmota.bin and tasmota.bin.gz: [pzem_dc_shuntconfig_1.zip](https://github.com/arendst/Tasmota/files/7614209/pzem_dc_shuntconfig_1.zip)
> 
> My development branch if you want to see the code: https://github.com/barbudor/Tasmota/blob/pzemdc_config_shunt/tasmota/xnrg_06_pzem_dc.ino#L140-L183

Hi, for me it doesnt work. But I have checked PZEM017 protocol and I have a way to configure it easily:

1. Configure tasmota as Serial Bridge (I have done it with hardware serial bridge but it must work also with softserial) (In my case I have soldered wires directly and "jump" the RS485 internal converter. I supposed that these commands could be send also throught rs485 converters)
2. Reboot
3. Configure serial: 
```
baudrate 9600
serialconfig 8N2
```
4. Send one of these commands (choose your option):
```
100A: serialsend5 01060003000079CA 
 50A: serialsend5 010600030001B80A
200A: serialsend5 010600030002F80B
300A: serialsend5 01060003000339CB
```
5. You will recieve an answer in serialrecieved. It must be just the same HEX string as you have send to PZEM017
6. Configure tasmota again as a PZEM017 and you will get the correct current and power values.

I hope you can fin this usefull

By the way: anyone? https://github.com/arendst/Tasmota/discussions/13733#discussioncomment-3620150

**Reply by freeesty** [reply]
if using softserial commands are: sbaudrate, sserialconfig 8N2, sserialsend5

**Comment by grappini** [comment]
I apologize for the late reply , i had a busy week. Unfortunately for the moment i cannot confirm it's working.
I had an incident with my pzem 017 while connecting it back to the system and now it's bricked , a really silly mistake.
I have several pzem's on their way from China , when i will receive one i'll get back with the results.

**Reply by barbudor** [reply]
Bad luck
Looking forward your feedback

**Reply by barbudor** [reply]
Any update on your PZEM from China ?

**Reply by grappini** [reply]
Still in transit. Seems that delivery from China In Europe is really slow on this time of year. I have some other items that are almost at the maximum waiting time before ask for refund. 
As soon as i get one pzem i will test it.

**Reply by barbudor** [reply]
Any update ?

**Reply by grappini** [reply]
I've just checked the tracking now , 2 pcs pzem 017 are in Europe at the airmail sorting hub , i hope to receive them soon.

**Reply by grappini** [reply]
Finally one pzem 017 arrived in my country , in a few days i should put my hands on it. It's annoying to see that a pzem 016 ac arrived from the same store in 3 weeks and the pzem's 017 took months.
The second pzem 017 is still at the airline hub in Europe.

**Reply by barbudor** [reply]
Ouch!
Most ALi orders comes in less than 3 weeks recently...
Now I have to read again to remind me what we were talking about 😂

**Comment by grappini** [comment]
The bad luck is continuing. There was 2 pcs pzem 017 ordered from China , one got lost on the way and the other is not functioning.
Unfortunately my medical condition took me away for a while from the "gadgets".
Today i finally tried the pzem 017 just to find out it's not working , defective from factory.
I also lost the right to open a dispute with the seller because i got to test it very very late.
But i don't give up , I'll order again , maybe i find a stock available in Europe

**Reply by pkkrusty** [reply]
I have a few coming in a week, will be able to test barbudor's work soon.

**Comment by admiralroflknife** [comment]
To revive this, I just finished wiring a esp32 to a pzem 17 for my solar setup today's the first day and I've noticed I'm getting basically double what I was expecting in terms of power Generation. This makes sense since I'm using a 50a shunt model, while the OP had half the output on the 200a model. The default config as it stands must be tuned to the 100a model. 

I'll be flashing barbudors file tonight and configure it to the 50a shunt and will report back tomorrow evening to see if it's successful. I'll probably also turn it off tomorrow evening and back on in the morning to see if that byte is saved to flash in the pzem or not.

**Reply by admiralroflknife** [reply]
I ended up flashing the firmware on finally (ended up switching to a esp8266 due to issues flashing it to my esp32), and installed it. I set up a little test. I had 80 watts feeding into my solar charge controller which pushed ~67 watts to my batteries through the shunt. In tasmota you can clearly see it is reading 135 watts which is exactly double of the real 67 watts.

I did end up sending the energyconfig command as show above, and ended up doing it for 50 all the way to 300 and the output stuck at 135 watts the whole time. Im guessing that the command to change it is not correctly getting to the pzem? 

Although I totally forgot to mention that I have done the mod to my pzem to remove the rs485 converter chip to have it directly connect to the pzem directly through serial (although I doubt thats the issue as all the rest of the data is coming through fine, realistically I think it should reduce the chance of it not getting through correctly).

I'm up to try any additional ideas or mods to the code if your willing to try. In the meantime I'll probably just modify my home assisstant code a little to just divide current and kwh by 2 for the time being. 

![Screenshot_20220510-213935](https://user-images.githubusercontent.com/101166534/167770933-5ddb97a8-86aa-438a-8b0e-e29a0fb9953f.png)
![PXL_20220511_043931634](https://user-images.githubusercontent.com/101166534/167770935-dff9be60-3ed4-472f-88af-ebed74d5c3c1.jpg)
![PXL_20220511_044129014](https://user-images.githubusercontent.com/101166534/167770937-5a92998f-a32e-4e01-a2e6-e50737e941e4.jpg)

**Comment by barbudor** [comment]
If you flashed my esp8266 binary to an ESP32 it is normal to have boot problem.
Would you prefer an ESP32 binary? 

Let me dive back into the topic, read your comments above and think about it.

**Reply by admiralroflknife** [reply]
Sounds good, I have both esp32 and esp8266 so it  doesn't really matter, thanks for working on it

**Reply by ortegafernando** [reply]
@admiralroflknife  Hi, any update in energyconfig  command? May be we could use a SerialSend. It will be used only "once", so may be it is better to write down the 4 different commands and.... done.

What do you think @barbudor ?

I have an PZEM017 and I could know HEX characters to send.

Two more questions:
1. how do you power your esp8266/esp32 ? Could we take a LDO connected to the 8.5v (more or less) rail in the PZEM? (you can see this voltage in the input of the little LDO 3.3v 30ma that the pzem has). How do you solve the power of the esp8266?
2. About direcction of the current flow... I have read that PZEM can measure it but it always sends "positive" values. In that case, we could take an resistor divider of the voltage that we are measuring, take a centered analog read and detect if it is over or under 0,5v (esp8266 has an analog input between 0-1v), we dont want accuracy, only now if it is over or under 0.5 (I am not an expert, may be it is better an oamp). With that value and the real value of the PZEM, for example in node red we could and "positive" or "negative" sign to the measurements. has it sense?  (of course, we couldn't use the energy values of pzem as they will be wrong if we have positive and negative currents)

Bye,

**Reply by barbudor** [reply]
I have''t had any feedback of the command so if you want to test it, I can rebuild an updated one for eithe 8266 or 32

As I expect the PZEM to store the shunt value, you may be right and it might be possible to configure it through SSendSerial before configuring the GPIO as PZEM and only after set as PZEM. However I don't see that as very handy.

On question 1. I don't know as I don't own a pzem17. But generally manufacturer don't oversize their power supplies so unsure that there will be enough remaining power for the ESP

On point 2. I don't get much your point (may be I'm too tired 😄 ) so may be a diagram?
Don't forget that adding a divider in parallel of any existing measurement system in the pz may disturb it.

**Reply by ortegafernando** [reply]
@barbudor  Thanks.

About 1. I have seen "an strange" power supply. It has an mosfet N-channel and a simply Zener diode. If input voltage is >12v then I get a 8.5v fixed after mosfet, if inout voltage is <12v is not 8.5v fixed, it decreases (see photo). I will try to get 3.3v or 5v from here with a new LDO. Of course I can NOT use the little 3.3v LDO becuase it only has a maximum output of 30mA. (I am not an expert, but I think that is a "simply zener voltage regulator"

![photo1662911868](https://user-images.githubusercontent.com/8700791/189537587-51a2ea37-4113-46ef-9efe-472421e9d5d7.jpeg)

2. Yes, you are rigth. I can't add any kind of resisitor in parallel. I am not an expert, but an high impedance device. Could I use an opamp? The main objetive here is only to dectect negative or positive voltage of the shunt (it can go from -75mV to 75mV). it doesn't matter the real value in mV. Only the "sign". So an opamp could be the best approach. I have also read about detecting negative voltages with an optocoupler

May be there is anyone here that knows about opamp, impedances, and so on.... 

ADDED: For example here I can find a solution:
https://www.ti.com/lit/ds/symlink/opa388.pdf?ts=1662872968696
Point 8.2.1 Bidirectional Current-Sensing
Do you think that with an input current bias of pA it will disturb the pzem circuit?

**Reply by ortegafernando** [reply]
> @barbudor Thanks.
> 
> About 1. I have seen "an strange" power supply. It has an mosfet N-channel and a simply Zener diode. If input voltage is >12v then I get a 8.5v fixed after mosfet, if inout voltage is <12v is not 8.5v fixed, it decreases (see photo). I will try to get 3.3v or 5v from here with a new LDO. Of course I can NOT use the little 3.3v LDO becuase it only has a maximum output of 30mA. (I am not an expert, but I think that is a "simply zener voltage regulator"
> 
> ![photo1662911868](https://user-images.githubusercontent.com/8700791/189537587-51a2ea37-4113-46ef-9efe-472421e9d5d7.jpeg)
> 
> 2. Yes, you are rigth. I can't add any kind of resisitor in parallel. I am not an expert, but an high impedance device. Could I use an opamp? The main objetive here is only to dectect negative or positive voltage of the shunt (it can go from -75mV to 75mV). it doesn't matter the real value in mV. Only the "sign". So an opamp could be the best approach. I have also read about detecting negative voltages with an optocoupler
> 
> May be there is anyone here that knows about opamp, impedances, and so on....
> 
> ADDED: For example here I can find a solution: https://www.ti.com/lit/ds/symlink/opa388.pdf?ts=1662872968696 Point 8.2.1 Bidirectional Current-Sensing Do you think that with an input current bias of pA it will disturb the pzem circuit?

Anyone?

**Comment by xury77** [comment]
Is it possible to get the direction of current flow through the pzem-017 shunt in tasmota?

**Comment by admiralroflknife** [comment]
As far as I’m aware no, it only adds to the value as it goes. It can’t
discern direction.

I got around this by using 2, one goes from the solar to the battery,
another going from battery to draw.

On Fri, Jul 15, 2022 at 7:00 AM xury77 ***@***.***> wrote:

> Is it possible to get the direction of current flow through the pzem-017
> shunt in tasmota?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/13733#discussioncomment-3157022>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYD23RXVXLBCRMTQRXGP2ZTVUFVHXANCNFSM5IM76APA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
--

**Comment by dexterbot80** [comment]
Hello, Can someone tell me what I'm doing wrong here?

I removed the MAX485 circuit and replaced it with two wires exactly as in the image below.
[https://user-images.githubusercontent.com/34340210/66046723-57efd780-e4f4-11e9-9e42-afc0a02392bb.jpeg](url)
[https://user-images.githubusercontent.com/34340210/66046724-57efd780-e4f4-11e9-9935-4e66b9154c17.jpeg](url)

Now I feed pzem 017 with 3.3V - I tried to feed with 3.3V from the D1mini module and I fed with 3.3V from a laboratory source same GND with D1mini
​

<img width="657" alt="image" src="https://github.com/arendst/Tasmota/assets/47325794/431a703f-bb83-428d-8bb2-7de23c879b8a">

<img width="1908" alt="image" src="https://github.com/arendst/Tasmota/assets/47325794/c6335063-694e-4334-8d4c-6203b570df2c">

**Reply by barbudor** [reply]
The photos shows a PZEM16 for AC
This discussion is about shunt settings on a PZEM17 for DC
Wrong place

**Reply by dexterbot80** [reply]
Hello, Thank you for all your work here. My problem was RX_TX connection 
I use PZEM 017 DC  and is working now :)

It would be very useful to post all the commands that can be given to the PZEM 017 module from the tasmota console.
Thank you @barbudor 
​

**Reply by dexterbot80** [reply]
Hello masters, Pzem 017 works good on ESP8266 but this device is to slow :) I would like to connect it to the ESP32 without the max485 circuit - can anyone come up with some pointers? Thank you

**Reply by barbudor** [reply]
[Here](https://github.com/arendst/Tasmota/discussions/19061#discussioncomment-6897408) aer some photos how I installed an ESP module directly inside my PZEM without cutting nor removing any parts.
It was an ESP-M3 (ESP8266) but could be done with any small ESP32 module like the SuperMini ESP32C3 or an ESP-C3-12

**Reply by dexterbot80** [reply]
I remove MAX485 chip and now PZEM 017 is connected to an ESP8266 - with this obsolte test code :). Indications on how to configure the ESP32 in this case and what commands to run would be useful. Thanks

**Reply by barbudor** [reply]
Thanks for reminding me that obsolte test code. I have now removed it as  the feature is part of Tasmota as per : https://github.com/arendst/Tasmota/pull/19402
Please remove your link

**Reply by dexterbot80** [reply]
And now i can use the PZEM017 DC Shunt without MAX485 IC ? For me simpler = more reliable

**Reply by barbudor** [reply]
Not sure to understand your question. Software changes has never been about using MAX485 or not but providing a way to set the shunt value from a Tasmota command.

Installing an ESP directly inside or removing and strapping the MAX485 is totally indépendant of the software version. As far as your setup provides a proper serial signal to the ESP.

**Reply by dexterbot80** [reply]
Yes, I'm interested in connecting the PZEM017 Shunt to ESP 32 without MAX485 and what commands I have to give it to work. Now I have installed Tasmota 13.4.0 seems to have caught my ears here :).  
```
09:00:05.942 CMD: EnergyConfig range=300
09:00:05.946 RSL: RESULT = {"EnergyConfig":"Done"}
09:00:22.954 CMD: teleperiod 10
09:00:22.958 RSL: RESULT = {"TelePeriod":10}
09:00:22.997 RSL: STATE = {"Time":"2024-03-04T09:00:22","Uptime":"0T00:01:18","UptimeSec":78,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"rdspro_IoT","BSSId":"F0:9F:C2:6A:72:27","Channel":11,"Mode":"11n","RSSI":100,"Signal":-50,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:00:23.027 RSL: SENSOR = {"Time":"2024-03-04T09:00:23","ENERGY":{"TotalStartTime":"2024-03-04T08:59:10","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"Voltage":0,"Current":0.000}}
09:00:32.979 RSL: STATE = {"Time":"2024-03-04T09:00:32","Uptime":"0T00:01:28","UptimeSec":88,"Heap":151,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"rdspro_IoT","BSSId":"F0:9F:C2:6A:72:27","Channel":11,"Mode":"11n","RSSI":100,"Signal":-47,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:00:33.008 RSL: SENSOR = {"Time":"2024-03-04T09:00:33","ENERGY":{"TotalStartTime":"2024-03-04T08:59:10","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"Voltage":0,"Current":0.000}}
09:00:42.972 RSL: STATE = {"Time":"2024-03-04T09:00:42","Uptime":"0T00:01:38","UptimeSec":98,"Heap":151,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"rdspro_IoT","BSSId":"F0:9F:C2:6A:72:27","Channel":11,"Mode":"11n","RSSI":100,"Signal":-43,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:00:43.003 RSL: SENSOR = {"Time":"2024-03-04T09:00:42","ENERGY":{"TotalStartTime":"2024-03-04T08:59:10","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"Voltage":0,"Current":0.000}}
09:00:46.176 CMD: baudrate 9600
09:00:46.179 SER: Set to 8N1 9600 bit/s
09:00:46.205 RSL: RESULT = {"Baudrate":9600}
09:00:52.978 RSL: STATE = {"Time":"2024-03-04T09:00:52","Uptime":"0T00:01:48","UptimeSec":108,"Heap":151,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"rdspro_IoT","BSSId":"F0:9F:C2:6A:72:27","Channel":11,"Mode":"11n","RSSI":100,"Signal":-46,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:00:53.233 RSL: SENSOR = {"Time":"2024-03-04T09:00:53","ENERGY":{"TotalStartTime":"2024-03-04T08:59:10","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"Voltage":0,"Current":0.000}}
09:01:02.951 RSL: STATE = {"Time":"2024-03-04T09:01:02","Uptime":"0T00:01:58","UptimeSec":118,"Heap":151,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"rdspro_IoT","BSSId":"F0:9F:C2:6A:72:27","Channel":11,"Mode":"11n","RSSI":100,"Signal":-49,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:01:03.204 RSL: SENSOR = {"Time":"2024-03-04T09:01:03","ENERGY":{"TotalStartTime":"2024-03-04T08:59:10","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"Voltage":0,"Current":0.000}}
09:01:04.212 CMD: serialconfig 8N2
09:01:04.215 SER: Set to 8N2 9600 bit/s
09:01:04.315 RSL: RESULT = {"SerialConfig":"8N2"}
09:01:12.972 RSL: STATE = {"Time":"2024-03-04T09:01:12","Uptime":"0T00:02:08","UptimeSec":128,"Heap":151,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"rdspro_IoT","BSSId":"F0:9F:C2:6A:72:27","Channel":11,"Mode":"11n","RSSI":100,"Signal":-42,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:01:13.251 RSL: SENSOR = {"Time":"2024-03-04T09:01:13","ENERGY":{"TotalStartTime":"2024-03-04T08:59:10","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"Voltage":0,"Current":0.000}}
09:01:22.952 RSL: STATE = {"Time":"2024-03-04T09:01:22","Uptime":"0T00:02:18","UptimeSec":138,"Heap":151,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"rdspro_IoT","BSSId":"F0:9F:C2:6A:72:27","Channel":11,"Mode":"11n","RSSI":100,"Signal":-41,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:01:23.237 RSL: SENSOR = {"Time":"2024-03-04T09:01:23","ENERGY":{"TotalStartTime":"2024-03-04T08:59:10","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"Voltage":0,"Current":0.000}}
```
![image](https://github.com/arendst/Tasmota/assets/47325794/153caf4e-838f-4d25-b6a8-932f88344fa0)
![image](https://github.com/arendst/Tasmota/assets/47325794/1be4afcf-2c20-463b-915e-fde1e8fe1470)


The shunt module is fed from the laboratory source with 13.8V now and in the Tasmota console I see 0V. The shunt resistor is not connected to the PZEM017 shunt module.

**Reply by barbudor** [reply]
Again, strapping the MAX485 has no impact on software
If it doesn't work, look on the hardware side
You may have not connected the ESP at the right spot

Note that the links to the photo and diagram I have posted above is about installing an ESP inside the PZEM, powered by the PZEM so it doesn't make use of the optocoupler

If you are using an ESP powered externaly, you need to conect on the MAX485 side (after the optocouplers) and strap some wires. I haven't seen any "official" way of strapping the MAX so take care of your connections.
You also need to remove any 120 ohms resistors accross A and B
I haven't analysed in details the schematics between the optocouplers and the external connector

**Reply by dexterbot80** [reply]
I opened the shunt case and remembered the existence of those two pull up resistors. If I fed with 3.3V it works now :) Initially I wanted to somehow receive power from the ether like Tesla :)
![WhatsApp Image 2024-03-04 at 14 07 23_fb1e58ea](https://github.com/arendst/Tasmota/assets/47325794/ad62d33f-328e-4579-a102-8882f29db389)

<img width="371" alt="image" src="https://github.com/arendst/Tasmota/assets/47325794/70992562-700d-49be-aa7e-b1ae0447ab87">

**Reply by barbudor** [reply]
Good to have that photo now as a reference

**Reply by dexterbot80** [reply]
Thank you for your help

**Comment by marine1988** [comment]
is possible to use tasmota with pzem 017 and rs485

**Comment by xury77** [comment]
Of course. If you connect transceiver rs485 to your Esp device

**Comment by marine1988** [comment]
Im using the esphome i like more then tasmota but since i need to set the shunt to 200 i installed the tasmota to do that but tru e rs485 is not reading the values and i cant send the settint to 200a shunt.

`11:43:14.379 CMD: serialconfig 8N2
11:43:14.385 MQT: stat/tasmota_78EF27/RESULT = {"SerialConfig":"8N2"}
11:43:28.849 CMD: baudrate 9600
11:43:28.854 MQT: stat/tasmota_78EF27/RESULT = {"Baudrate":9600}
11:43:52.235 CMD: serialsend5 01 03 00 00 00 04 44 09
11:43:52.240 MQT: stat/tasmota_78EF27/RESULT = {"SerialSend":"Done"}`

Here is my config connection my pzem is stock I don't solder anything to it


![20240701_104441](https://github.com/arendst/Tasmota/assets/6524068/d9a00fe4-5055-451d-b316-989b3b8d9de5)
![20240701_104436](https://github.com/arendst/Tasmota/assets/6524068/ab19d037-9904-4fcc-9f92-d5ef0bb8cc1e)

**Reply by barbudor** [reply]
Have you read this thread correctly ?
There is a dedicated command to set the shunt as described [here](#discussioncomment-1712501)

**Comment by freeesty** [comment]
I've got a new pzem 017 v2, which I wanted to use directly to esp32c3, without modbus. But the current reading is off somehow.
I've already remove the rs485 chip and flashed tasmota, tried the shunt commands above for 50A shunt, but still - only voltage works, but amps are 5x too low.

**Reply by barbudor** [reply]
Have you connected the shunt correctly ?
From the photo below, which one shows your shunt ?

<img width="900" height="900" alt="17828456839193729442573511859807" src="https://github.com/user-attachments/assets/7d02297b-e19f-4d02-8e94-0724690361b9" />

**Reply by freeesty** [reply]
thx for your reply.
Its #3
and connected properly I guess. Its my second pzem, but the first is with modbus.

<img width="4000" height="2252" alt="20260630_214837" src="https://github.com/user-attachments/assets/0db94d70-3a44-4df3-8ae8-ed0f50689b9f" />

**Reply by barbudor** [reply]
Indeed, seems to be the 50A one

Unless the shunt is damaged ...

No more idea

**Reply by freeesty** [reply]
It is all good with the shunt - cross test it with an other module and the orig. software in windows.

Then had another look at the commands, and it must be sth wrong with my softserial commands. Did it again and its working now.

when using softserial bridge its:
SBaudrate 9600
SSerialConfig 8N2
SSerialSend5 01060003000079CA (100A shunt) or corrospending from this comment https://github.com/arendst/Tasmota/discussions/13733#discussioncomment-3827871
